app/src/main/java/app/revanced/integrations/patches/playback/speed/CustomVideoSpeedPatch.java

```diff
@@ -1,12 +1,80 @@
 package app.revanced.integrations.patches.playback.speed;
 
+import androidx.annotation.NonNull;
+
+import java.util.Arrays;
+
+import app.revanced.integrations.settings.SettingsEnum;
+import app.revanced.integrations.utils.LogHelper;
+import app.revanced.integrations.utils.ReVancedUtils;
+
 public class CustomVideoSpeedPatch {
     /**
-     * Default playback speeds offered by YouTube.
-     * Values are also used by {@link RememberPlaybackSpeedPatch}.
+     * Maximum playback speed, exclusive value.  Custom speeds must be less than this value.
+     */
+    public static final float MAXIMUM_PLAYBACK_SPEED = 10;
+
+    /**
+     * Custom playback speeds.
      *
-     * If custom video speed is applied,
-     * then this array is overwritten by the patch with custom speeds
+     * Values are also used by {@link RememberPlaybackSpeedPatch}.
+     */
+    public static float[] videoSpeeds;
+
+    /**
+     * Minimum value of {@link #videoSpeeds}
+     */
+    public static float minVideoSpeed;
+
+    /**
+     * Maxium value of {@link #videoSpeeds}
      */
-    public static final float[] videoSpeeds = {0.25f, 0.75f, 1.0f, 1.25f, 1.5f, 1.75f, 2.0f};
+    public static float maxVideoSpeed;
+
+    static {
+        loadSpeeds();
+    }
+
+    private static void resetCustomSpeeds(@NonNull String toastMessage) {
+        ReVancedUtils.showToastLong(toastMessage);
+        SettingsEnum.PLAYBACK_SPEED_AVAILABLE.saveValue(SettingsEnum.PLAYBACK_SPEED_AVAILABLE.defaultValue);
+    }
+
+    private static void loadSpeeds() {
+        try {
+            String[] speedStrings = SettingsEnum.PLAYBACK_SPEED_AVAILABLE.getString()
+                    .replaceAll("\\s", "").split(",");
+            Arrays.sort(speedStrings);
+            if (speedStrings.length == 0) {
+                throw new IllegalArgumentException();
+            }
+            videoSpeeds = new float[speedStrings.length];
+            for (int i = 0, length = speedStrings.length; i < length; i++) {
+                final float speed = Float.parseFloat(speedStrings[i]);
+                if (speed <= 0 || arrayContains(videoSpeeds, speed)) {
+                    throw new IllegalArgumentException();
+                }
+                if (speed >= MAXIMUM_PLAYBACK_SPEED) {
+                    resetCustomSpeeds("Custom speeds must be less than " + MAXIMUM_PLAYBACK_SPEED
+                            + ".  Using default values.");
+                    loadSpeeds();
+                    return;
+                }
+                minVideoSpeed = Math.min(minVideoSpeed, speed);
+                maxVideoSpeed = Math.max(maxVideoSpeed, speed);
+                videoSpeeds[i] = speed;
+            }
+        } catch (Exception ex) {
+            LogHelper.printInfo(() -> "parse error", ex);
+            resetCustomSpeeds("Invalid custom video speeds. Using default values.");
+            loadSpeeds();
+        }
+    }
+
+    private static boolean arrayContains(float[] array, float value) {
```

Comment:
Iirc this method is a duplicate of another somewhere else and should instead be refactored to the existing utils class.